crates/nu-command/src/filesystem/cp.rs

```diff
@@ -372,239 +94,141 @@ impl Command for Cp {
+                None,
+                None,
+                Vec::new(),
+            ));
         }
-
-        // Read src file
-        match buf_reader.read(&mut buffer) {
-            // src file read successfully
-            Ok(bytes_read) => {
-                // Write dst file
-                match buf_writer.write(&buffer[..bytes_read]) {
-                    // dst file written successfully
-                    Ok(bytes_written) => {
-                        // Update the total amount of bytes that has been saved and then print the progress bar
-                        bytes_processed += bytes_written as u64;
-                        bar.update_bar(bytes_processed);
-
-                        // the last block of bytes is going to be lower than the buffer size
-                        // let's break the loop once we write the last block
-                        if bytes_read < buffer.len() {
-                            break;
+        let target = paths.pop().expect("Should not be reached?");
+        let target_path = PathBuf::from(&target.item);
+        if target.item.ends_with('/') && !target_path.is_dir() {
+            return Err(ShellError::GenericError(
+                "is not a directory".into(),
+                "is not a directory".into(),
+                Some(target.span),
+                None,
+                Vec::new(),
+            ));
+        };
+        // paths now contains the sources
+        let sources: Vec<Vec<PathBuf>> = paths
+            .iter()
+            .map(|p| {
+                // Need to expand too make it work with globbing
+                let expanded_src = expand_to_real_path(&p.item);
+                match nu_glob::glob_with(&expanded_src.to_string_lossy(), GLOB_PARAMS) {
+                    Ok(files) => {
+                        let f = files.filter_map(Result::ok).collect::<Vec<PathBuf>>();
+                        if f.is_empty() {
+                            return Err(ShellError::FileNotFound(p.span));
                         }
+                        Ok(f)
                     }
-                    Err(e) => {
-                        // There was a problem writing the dst file
-                        process_failed = Some(e);
-                        break;
-                    }
+                    Err(e) => Err(ShellError::GenericError(
+                        e.to_string(),
+                        "invalid pattern".to_string(),
+                        Some(p.span),
+                        None,
+                        Vec::new(),
+                    )),
                 }
-            }
-            Err(e) => {
-                // There was a problem reading the src file
-                process_failed = Some(e);
-                break;
-            }
+            })
+            .collect::<Result<Vec<Vec<PathBuf>>, ShellError>>()?;
+
+        let sources = sources.into_iter().flatten().collect::<Vec<PathBuf>>();
+        let options = uu_cp::Options {
```

Comment:
It might be nice to reorder this into:
```rust
let options = uu_cp::Options {
    overwrite,
    reflink_mode,
    // etc.
    
    attributes_only: false,
    backup: BackupMode::NoBackup,
    // etc.
};
```
This might make it a bit easier to see what fields you're setting and which ones are just constant values. I also wonder if uutils could provide `Options::default` 🤔 Although, I guess that's not very important.

Reply:
Hmm I really like this idea, so it could end up looking like (for us) :
```
Options {
            overwrite,
            reflink_mode,
            recursive,
            debug,
            verbose: verbose || debug,
            dereference: !recursive,
            progress_bar: progress,
            ..
}
```